File: smql-engine/crates/smql-ast/src/expression.rs
use serde::{Deserialize, Serialize};
use std::fmt;

use crate::span::Span;
use crate::value::{SmqlDuration, Value};

/// Expressions used in guards, WHERE clauses, and MUTATE clauses.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Expression {
    pub kind: ExpressionKind,
    pub span: Option<Span>,
}

impl Expression {
    pub fn new(kind: ExpressionKind) -> Self {
        Self { kind, span: None }
    }

    pub fn with_span(kind: ExpressionKind, span: Span) -> Self {
        Self {
            kind,
            span: Some(span),
        }
    }
}

/// The core expression types.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum ExpressionKind {
    // Literals
    Literal(Value),
    DurationLiteral(SmqlDuration),

    // Field access
    /// Simple field reference: `priority`, `assignee`
    FieldAccess(Vec<String>),
    /// SELF keyword — refers to current instance
    SelfRef,
    /// ACTOR keyword — refers to the actor performing the transition
    ActorRef,
    /// ACTOR.field or SELF.field with path
    QualifiedAccess {
        root: Box<Expression>,
        path: Vec<String>,
    },

    // Binary operations
    BinaryOp {
        left: Box<Expression>,
        op: BinaryOperator,
        right: Box<Expression>,
    },

    // Unary operations
    UnaryOp {
        op: UnaryOperator,
        operand: Box<Expression>,
    },

    // Function calls
    FunctionCall {
        name: String,
        args: Vec<Expression>,
    },

    // State predicates
    /// STATE IS <state_name>
    StateIs(String),
    /// STATE IN { state1, state2, ... }
    StateIn(Vec<String>),

    // Null checks
    /// field IS SET
    IsSet(Box<Expression>),
    /// field IS NOT SET / IS NULL
    IsNotSet(Box<Expression>),

    // Set membership
    /// expr IN { val1, val2, ... }
    InSet {
        expr: Box<Expression>,
        values: Vec<Expression>,
    },
    /// expr IN (string_val1, string_val2, ...)
    InList {
        expr: Box<Expression>,
        values: Vec<Expression>,
    },

    // Collection predicates
    /// ALL(collection, predicate)
    All {
        collection: Box<Expression>,
        predicate: Box<Expression>,
    },
    /// ANY(collection, predicate)
    Any {
        collection: Box<Expression>,
        predicate: Box<Expression>,
    },
    /// COUNT(collection) or COUNT()
    Count(Option<Box<Expression>>),

    // Signal predicate (for transition guards)
    /// SIGNAL FROM Machine WHERE condition
    SignalFrom {
        machine: String,
        condition: Box<Expression>,
    },

    // Pattern matching
    /// PATTERN(regex_string)
    Pattern(String),
}

/// Binary operators.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BinaryOperator {
    // Comparison
    Eq,
    NotEq,
    Lt,
    Gt,
    LtEq,
    GtEq,
    // Arithmetic
    Add,
    Sub,
    Mul,
    Div,
    // Logical
    And,
    Or,
}

impl fmt::Display for BinaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BinaryOperator::Eq => write!(f, "=="),
            BinaryOperator::NotEq => write!(f, "!="),
            BinaryOperator::Lt => write!(f, "<"),
            BinaryOperator::Gt => write!(f, ">"),
            BinaryOperator::LtEq => write!(f, "<="),
            BinaryOperator::GtEq => write!(f, ">="),
            BinaryOperator::Add => write!(f, "+"),
            BinaryOperator::Sub => write!(f, "-"),
            BinaryOperator::Mul => write!(f, "*"),
            BinaryOperator::Div => write!(f, "/"),
            BinaryOperator::And => write!(f, "AND"),
            BinaryOperator::Or => write!(f, "OR"),
        }
    }
}

/// Unary operators.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum UnaryOperator {
    Not,
    Neg,
}

impl fmt::Display for UnaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UnaryOperator::Not => write!(f, "NOT"),
            UnaryOperator::Neg => write!(f, "-"),
        }
    }
}

impl fmt::Display for Expression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.kind {
            ExpressionKind::Literal(v) => write!(f, "{}", v),
            ExpressionKind::DurationLiteral(d) => write!(f, "{}", d),
            ExpressionKind::FieldAccess(path) => write!(f, "{}", path.join(".")),
            ExpressionKind::SelfRef => write!(f, "SELF"),
            ExpressionKind::ActorRef => write!(f, "ACTOR"),
            ExpressionKind::QualifiedAccess { root, path } => {
                write!(f, "{}.{}", root, path.join("."))
            }
            ExpressionKind::BinaryOp { left, op, right } => {
                write!(f, "({} {} {})", left, op, right)
            }
            ExpressionKind::UnaryOp { op, operand } => write!(f, "({} {})", op, operand),
            ExpressionKind::FunctionCall { name, args } => {
                let arg_strs: Vec<String> = args.iter().map(|a| a.to_string()).collect();
                write!(f, "{}({})", name, arg_strs.join(", "))
            }
            ExpressionKind::StateIs(state) => write!(f, "STATE IS {}", state),
            ExpressionKind::StateIn(states) => write!(f, "STATE IN {{{}}}", states.join(", ")),
            ExpressionKind::IsSet(expr) => write!(f, "{} IS SET", expr),
            ExpressionKind::IsNotSet(expr) => write!(f, "{} IS NOT SET", expr),
            ExpressionKind::InSet { expr, values } => {
                let val_strs: Vec<String> = values.iter().map(|v| v.to_string()).collect();
                write!(f, "{} IN {{{}}}", expr, val_strs.join(", "))
            }
            ExpressionKind::InList { expr, values } => {
                let val_strs: Vec<String> = values.iter().map(|v| v.to_string()).collect();
                write!(f, "{} IN ({})", expr, val_strs.join(", "))
            }
            ExpressionKind::All {
                collection,
                predicate,
            } => write!(f, "ALL({}, {})", collection, predicate),
            ExpressionKind::Any {
                collection,
                predicate,
            } => write!(f, "ANY({}, {})", collection, predicate),
            ExpressionKind::Count(expr) => match expr {
                Some(e) => write!(f, "COUNT({})", e),
                None => write!(f, "COUNT()"),
            },
            ExpressionKind::SignalFrom { machine, condition } => {
                write!(f, "SIGNAL FROM {} WHERE {}", machine, condition)
            }
            ExpressionKind::Pattern(regex) => write!(f, "PATTERN({})", regex),
        }
    }
}

File: sdk-typescript/src/builder/expression.ts
import { valueToSmql } from "./value.js";
import type { SmqlValue } from "../types.js";

export class Expr {
  private readonly expr: string;

  private constructor(expr: string) {
    this.expr = expr;
  }

  toString(): string {
    return this.expr;
  }

  // --- Factories ---

  static field(name: string): Expr {
    return new Expr(name);
  }

  static val(v: SmqlValue): Expr {
    return new Expr(valueToSmql(v));
  }

  static stateIs(state: string): Expr {
    return new Expr(`STATE IS ${state}`);
  }

  static stateIn(...states: string[]): Expr {
    return new Expr(`STATE IN { ${states.join(", ")} }`);
  }

  static isSet(field: string): Expr {
    return new Expr(`${field} IS SET`);
  }

  static isNotSet(field: string): Expr {
    return new Expr(`${field} IS NOT SET`);
  }

  static raw(expr: string): Expr {
    return new Expr(expr);
  }

  // --- Query predicates ---

  static alive(): Expr {
    return new Expr("ALIVE");
  }

  static terminated(): Expr {
    return new Expr("TERMINATED");
  }

  static stuckIn(state: string, duration: string): Expr {
    return new Expr(`STUCK_IN("${state}", ${duration})`);
  }

  static hasVisited(state: string): Expr {
    return new Expr(`HAS_VISITED("${state}")`);
  }

  static neverVisited(state: string): Expr {
    return new Expr(`NEVER_VISITED("${state}")`);
  }

  static tag(key: string, value: string): Expr {
    return new Expr(`TAG "${key}" == "${value}"`);
  }

  // --- Composition predicates ---

  static parentState(): Expr {
    return new Expr("PARENT.STATE");
  }

  static parentField(field: string): Expr {
    return new Expr(`PARENT.${field}`);
  }

  static signalFrom(machine: string, condition: string): Expr {
    return new Expr(`SIGNAL FROM ${machine} WHERE ${condition}`);
  }

  static all(collection: string, predicate: string): Expr {
    return new Expr(`ALL(${collection}, ${predicate})`);
  }

  static any(collection: string, predicate: string): Expr {
    return new Expr(`ANY(${collection}, ${predicate})`);
  }

  static countOf(collection: string): Expr {
    return new Expr(`COUNT(${collection})`);
  }

  // --- Built-in functions ---

  static elapsed(): Expr {
    return new Expr("elapsed()");
  }

  static elapsedSince(state: string): Expr {
    return new Expr(`elapsed_since("${state}")`);
  }

  static now(): Expr {
    return new Expr("NOW()");
  }

  static today(): Expr {
    return new Expr("TODAY()");
  }

  static timeoutRemaining(): Expr {
    return new Expr("timeout_remaining()");
  }

  static len(expr: string): Expr {
    return new Expr(`len(${expr})`);
  }

  static lower(expr: string): Expr {
    return new Expr(`lower(${expr})`);
  }

  static upper(expr: string): Expr {
    return new Expr(`upper(${expr})`);
  }

  static pattern(regex: string): Expr {
    return new Expr(`PATTERN("${regex}")`);
  }

  // --- Comparisons ---

  eq(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} == ${toExprStr(other)}`);
  }

  neq(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} != ${toExprStr(other)}`);
  }

  gt(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} > ${toExprStr(other)}`);
  }

  gte(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} >= ${toExprStr(other)}`);
  }

  lt(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} < ${toExprStr(other)}`);
  }

  lte(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} <= ${toExprStr(other)}`);
  }

  // --- Logical ---

  and(other: Expr): Expr {
    return new Expr(`(${this.expr}) AND (${other.expr})`);
  }

  or(other: Expr): Expr {
    return new Expr(`(${this.expr}) OR (${other.expr})`);
  }

  not(): Expr {
    return new Expr(`NOT (${this.expr})`);
  }

  // --- Arithmetic ---

  add(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} + ${toExprStr(other)}`);
  }

  sub(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} - ${toExprStr(other)}`);
  }

  mul(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} * ${toExprStr(other)}`);
  }

  div(other: Expr | SmqlValue): Expr {
    return new Expr(`${this.expr} / ${toExprStr(other)}`);
  }

  // --- Field access ---

  dot(field: string): Expr {
    return new Expr(`${this.expr}.${field}`);
  }

  // --- Set membership ---

  in(...values: SmqlValue[]): Expr {
    const items = values.map(valueToSmql);
    return new Expr(`${this.expr} IN { ${items.join(", ")} }`);
  }
}

function toExprStr(v: Expr | SmqlValue): string {
  if (v instanceof Expr) return v.toString();
  return valueToSmql(v);
}
